compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -429,14 +430,36 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
     }
 
     fn before_terminator(ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // The step limit has already been hit in a previous call to `before_terminator`.
-        if ecx.machine.steps_remaining == 0 {
-            return Ok(());
-        }
-
-        ecx.machine.steps_remaining -= 1;
-        if ecx.machine.steps_remaining == 0 {
-            throw_exhaust!(StepLimitReached)
+        if let Some(new_steps) = ecx.machine.num_evaluated_terminators.checked_add(1) {
+            ecx.machine.num_evaluated_terminators = new_steps;
+            // Only report after a certain number of terminators have been evaluated and the
+            // current number of evaluated terminators is a power of 2. The latter gives us a cheap
+            // way to implement exponential backoff.
+            if new_steps > 1_000_000 && new_steps.count_ones() == 1 {
```

Comment:
```suggestion
            if new_steps > 1_000_000 && new_steps.is_power_of_two() {
```
That should be just as efficient and a little nicer